akka-http-core/src/main/scala/akka/http/impl/engine/client/ProxyGraphStage.scala

```diff
@@ -0,0 +1,119 @@
+import akka.http.scaladsl.settings.ClientConnectionSettings
+import akka.stream.stage.{ GraphStage, GraphStageLogic, InHandler, OutHandler }
+import akka.stream.{ Attributes, BidiShape, Inlet, Outlet }
+import akka.util.ByteString
+
+object ProxyGraphStage {
+  sealed trait State
+  // Entry state
+  case object Starting extends State
+
+  // State after CONNECT messages has been sent to Proxy and before Proxy responded back
+  case object Connecting extends State
+
+  // State after Proxy responded  back
+  case object Connected extends State
+}
+
+class ProxyGraphStage(targetHostName: String, targetPort: Int, settings: ClientConnectionSettings, log: LoggingAdapter)
+  extends GraphStage[BidiShape[ByteString, ByteString, ByteString, ByteString]] {
+
+  import ProxyGraphStage._
+
+  val bytesIn: Inlet[ByteString] = Inlet("OutgoingTCP.in")
+  val bytesOut: Outlet[ByteString] = Outlet("OutgoingTCP.out")
+
+  val sslIn: Inlet[ByteString] = Inlet("OutgoingSSL.in")
+  val sslOut: Outlet[ByteString] = Outlet("OutgoingSSL.out")
+
+  override def shape: BidiShape[ByteString, ByteString, ByteString, ByteString] = BidiShape.apply(sslIn, bytesOut, bytesIn, sslOut)
+
+  private val connectMsg = ByteString(s"CONNECT ${targetHostName}:${targetPort} HTTP/1.1\r\nHost: ${targetHostName}\r\n\r\n")
+
+  @scala.throws[Exception](classOf[Exception])
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+    private var state: State = Starting
+
+    val parser = new HttpResponseParser(settings.parserSettings, HttpHeaderParser(settings.parserSettings, log)()) {
+      var first = true
+      override def handleInformationalResponses = false
+      override protected def parseMessage(input: ByteString, offset: Int): StateResult = {
+        if (first) {
+          first = false
+          super.parseMessage(input, offset)
+        } else {
+          emit(RemainingBytes(input.drop(offset)))
+          terminate()
+        }
+      }
+    }
+    parser.setContextForNextResponse(HttpResponseParser.ResponseContext(HttpMethods.CONNECT, None))
+
+    setHandler(sslIn, new InHandler {
+      override def onPush() = {
+        state match {
+          case Starting   ⇒
+          // should not happen
+          case Connecting ⇒
+          // should not happen
+          case Connected ⇒
+            push(bytesOut, grab(sslIn))
+        }
+      }
+    })
+
+    setHandler(bytesIn, new InHandler {
+      override def onPush() = {
+        state match {
+          case Starting ⇒
+          // should not happen to be in Starting state
+          case Connecting ⇒
+            val proxyResponse = grab(bytesIn)
+            parser.parseBytes(proxyResponse) match {
+              case NeedMoreData ⇒ pull(bytesIn)
+              case ResponseStart(StatusCodes.OK, _, _, _, _) ⇒
+                state = Connected
+                pull(sslIn)
+                pull(bytesIn)
```

Comment:
That one is little bit problematic, we're touching a few things here:

1) We are not sending anything until 2xx response come back from server. According to RFC (https://tools.ietf.org/html/rfc2817#section-5.3) we can send content to be sent to target host right after CONNECT. That may result in overall faster response but wanted to start with simple approach. Is it ok?
2) If we stick to current approach to 1) then I see little chance of server sending us anything more than 2xx response and even if that happens I am not sure how to treat this. As we haven't even sent request it's hard to interpret this as a response.
3) It's more like an observation that may be useful if we want to be more diligent regarding parsing here. Actually we are not getting `MessageEnd` at all here. I guess that is because connection is not closed and there's not other mean for Parser to know that the 2xx message has ended. So for the first `pull` we are getting either `NeedMoreData` (for manual testing with Charles proxy or  `flowOutProbe.sendNext(ByteString("HTTP/1.0 200 Connection established\r\n\r\n"))` as in https://github.com/akka/akka-http/pull/921/files#diff-f2ab7fa3256d9d613bdc4d73f5479281R32) or even `EntityPart` if I code proxy response as `flowOutProbe.sendNext(ByteString("HTTP/1.0 200 Connection established\r\n\r\nSomethingMore))"` in test.